_drafts/zero-to-main-libc.md

```diff
@@ -0,0 +1,721 @@
+---
+title: "From zero to main(): Bootstrapping libc with Newlib"
+description:
+  "TODO"
+author: francois
+tags: [zero-to-main]
+---
+
+In this series of post, we’ve worked methodically to demystify what happens to
+firmware before the main() function is called. So far, we bootstrapped a C
+environment, wrote a linker script from scratch, and implemented our own
+bootloader.
+
+And yet, we cannot even write a hello world program! Consider the following
+`main.c` file:
+
+```c
+#include <stdio.h>
+
+int main() {
+  printf("Hello, World\n");
+  while (1) {}
+}
+```
+Compiling this using our makefile and linker script from [previous](% post_url
+2019-05-14-zero-to-main-1 %) [posts](% post_url
+2019-06-25-how-to-write-linker-scripts-for-firmware %), we hit
+the following error:
+
+```shell
+$ make
+...
+Linking build/minimal.elf
+/usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/gcc/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: build/objs/a/b/c/minimal.o: in function `main':
+/minimal/minimal.c:4: undefined reference to `printf'
+collect2: error: ld returned 1 exit status
+make: *** [build/minimal.elf] Error 1
+```
+
+Undefined reference to `printf`! How could this be? Our firmware’s C environment
+is still missing a key component: a working C standard library.  This means that
```

Comment:
I'd love to have `working C standard library` bolded, but that's me

Reply:
🙅‍♂